### PR TITLE
MAINT: malloc return value checks for cython

### PR DESCRIPTION
**Pull request by sethtroisi**
#### Reference issue

Starting work on https://github.com/scipy/scipy/issues/9999
Taking over https://github.com/scipy/scipy/pull/10035/files

### CODE REVIEW

File: scipy/interpolate/_ppoly.pyx [hunk 1]
```diff
@@ -458,6 +458,11 @@ def real_roots(double[:,:,::1] c, double[::1] x, double y, bint report_discont,
 
     wr = <double*>libc.stdlib.malloc(c.shape[0] * sizeof(double))
     wi = <double*>libc.stdlib.malloc(c.shape[0] * sizeof(double))
+    if not wr or not wi:
+        libc.stdlib.free(wr)
+        libc.stdlib.free(wr)
+        raise MemoryError("Failed to allocate memory in real_roots")
```
Comment (sethtroisi) [inline scipy/interpolate/_ppoly.pyx:464]: @WarrenWeckesser You left comments in https://github.com/scipy/scipy/pull/10035/files#r279160979 about `raise` not working in `cdef` without `except -1`. I read you link and the way I processed it is that `def` functions are pure python and I'm allowed to `raise` like normal in them.

I believe I have correctly added `except -1` or similar where needed adding a return int to some functions. I'm still new to cython so a more experienced pair of eyes would be appreciated. Thanks

File: scipy/interpolate/_ppoly.pyx [hunk 2]
```diff
@@ -796,7 +800,7 @@ cdef double_or_complex evaluate_poly1(double s, double_or_complex[:,:,::1] c, in
 @cython.boundscheck(False)
 @cython.cdivision(True)
 cdef int croots_poly1(double[:,:,::1] c, double y, int ci, int cj,
-                      double* wr, double* wi, void **workspace):
+                      double* wr, double* wi, void **workspace) except -10:
```
Comment (sethtroisi) [inline scipy/interpolate/_ppoly.pyx:803]: -1 and -2 are valid return values currently in use.

File: scipy/spatial/setlist.pxd [hunk 3]
```diff
@@ -47,6 +47,14 @@ cdef inline int init(setlist_t *setlist, size_t n, size_t size_guess) except -1:
         setlist.sizes[j] = 0
         setlist.alloc_sizes[j] = size_guess
         setlist.sets[j] = <int*>libc.stdlib.malloc(sizeof(int) * size_guess)
+        if setlist.alloc_sizes == NULL:
+            # free all sets[i]
+            for i in range(j):
+                libc.stdlib.free(<void*>setlist.sets[i])
```
Comment (sethtroisi) [inline scipy/spatial/setlist.pxd:53]: This is almost a direct copy of `free()` directly after this.
it feels weird to pass `<void*>` but it's what `free()` does so I copied assuming I don't understand something.

**Review by sethtroisi — COMMENTED** [review]
I'm not sure how to test this code.
I can probably set `ulimit` then call python and try to solve a large problem and see what happens?
I'm open to other advice or suggestions

File: scipy/interpolate/_ppoly.pyx [hunk 4]
```diff
@@ -458,6 +458,11 @@ def real_roots(double[:,:,::1] c, double[::1] x, double y, bint report_discont,
 
     wr = <double*>libc.stdlib.malloc(c.shape[0] * sizeof(double))
     wi = <double*>libc.stdlib.malloc(c.shape[0] * sizeof(double))
+    if not wr or not wi:
+        libc.stdlib.free(wr)
```
Comment (ilayn) [inline scipy/interpolate/_ppoly.pyx:462]: This one should read `wi` I think.

File: scipy/spatial/setlist.pxd [hunk 5]
```diff
@@ -47,6 +47,14 @@ cdef inline int init(setlist_t *setlist, size_t n, size_t size_guess) except -1:
         setlist.sizes[j] = 0
         setlist.alloc_sizes[j] = size_guess
         setlist.sets[j] = <int*>libc.stdlib.malloc(sizeof(int) * size_guess)
+        if setlist.alloc_sizes == NULL:
```
Comment (pv) [inline scipy/spatial/setlist.pxd:50]:
```suggestion
        if setlist.sets[j] == NULL:
```

File: scipy/interpolate/_ppoly.pyx (hunk 4 above)
Comment (sethtroisi) [reply scipy/interpolate/_ppoly.pyx:462]: Thank you for reading. My eyes just glazed over this as I was reviewing.

File: scipy/spatial/setlist.pxd (hunk 5 above)
Comment (sethtroisi) [reply scipy/spatial/setlist.pxd]: Thanks

**Comment by sethtroisi** [comment]
Ping @WarrenWeckesser I'd appreciate your review of this

**Comment by WarrenWeckesser** [comment]
@sethtroisi, thanks for working on this.  I don't think I'll time to review this until later in the week.

File: scipy/spatial/setlist.pxd [hunk 6]
```diff
@@ -47,6 +47,13 @@ cdef inline int init(setlist_t *setlist, size_t n, size_t size_guess) except -1:
         setlist.sizes[j] = 0
         setlist.alloc_sizes[j] = size_guess
         setlist.sets[j] = <int*>libc.stdlib.malloc(sizeof(int) * size_guess)
+        if setlist.sets[j] == NULL:
+            for i in range(j):
+                libc.stdlib.free(<void*>setlist.sets[i])
+            libc.stdlib.free(<void*>setlist.sets)
+            libc.stdlib.free(<void*>setlist.sizes)
+            raise MemoryError
+
```
Comment (WarrenWeckesser) [inline scipy/spatial/setlist.pxd:56]: Remove this extra blank line.
```suggestion
```

File: scipy/spatial/setlist.pxd (hunk 3 above)
Comment (WarrenWeckesser) [reply scipy/spatial/setlist.pxd:52]: No problem.  It probably isn't necessary, but maybe some compilers complain if it isn't used?

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
@sethtroisi, thanks, this looks good.  I requested a few small changes.

File: scipy/spatial/setlist.pxd [hunk 7]
```diff
@@ -47,6 +47,12 @@ cdef inline int init(setlist_t *setlist, size_t n, size_t size_guess) except -1:
         setlist.sizes[j] = 0
         setlist.alloc_sizes[j] = size_guess
         setlist.sets[j] = <int*>libc.stdlib.malloc(sizeof(int) * size_guess)
+        if setlist.sets[j] == NULL:
+            for i in range(j):
+                libc.stdlib.free(<void*>setlist.sets[i])
+            libc.stdlib.free(<void*>setlist.sets)
+            libc.stdlib.free(<void*>setlist.sizes)
```
Comment (WarrenWeckesser) [inline scipy/spatial/setlist.pxd:54]: Looks like I forgot to click on the "add comment" button for one of my comments in my last review.  We also need to free `setlist.alloc_sizes` here.
```suggestion
            libc.stdlib.free(<void*>setlist.sizes)
            libc.stdlib.free(<void*>setlist.alloc_sizes)
```

**Review by WarrenWeckesser — APPROVED** [review]

**Comment by WarrenWeckesser** [comment]
Merged.  Thanks @sethtroisi.

**Pull request merged**